### PR TITLE
Replace tuples by classes in asyncoperation_sendmany

### PR DESCRIPTION
**Pull request by oxarbitrage**
Close https://github.com/zcash/zcash/issues/1361

### CODE REVIEW

File: src/wallet/asyncrpcoperation_sendmany.h [hunk 1]
```diff
@@ -25,14 +25,49 @@
...
+    std::string address;
+    CAmount amount;
+    std::string memo;
+
+    SendManyRecipient(std::string address_, CAmount amount_, std::string memo_ = "")
+    {
+        address = address_;
+        amount = amount_;
+        memo = memo_;
+    }
```
Comment (str4d) [inline src/wallet/asyncrpcoperation_sendmany.h]:
```suggestion
    SendManyRecipient(std::string address_, CAmount amount_, std::string memo_ = "") :
        address(address_), amount(amount_), memo(memo_) {}
```

File: src/wallet/asyncrpcoperation_sendmany.h [hunk 2]
```diff
@@ -25,14 +25,49 @@
...
 
-// Input JSOP is a tuple of JSOutpoint, note and amount
-typedef std::tuple<JSOutPoint, SproutNote, CAmount> SendManyInputJSOP;
+    SendManyInputUTXO(uint256 txid_, int vout_, CAmount amount_, bool coinbase_)
+    {
+        txid = txid_;
+        vout = vout_;
+        amount = amount_;
+        coinbase = coinbase_;
+    }
```
Comment (str4d) [inline src/wallet/asyncrpcoperation_sendmany.h]:
```suggestion
    SendManyInputUTXO(uint256 txid_, int vout_, CAmount amount_, bool coinbase_) :
        txid(txid_), vout(vout_), amount(amount_), coinbase(coinbase_) {}
```

File: src/wallet/asyncrpcoperation_sendmany.h [hunk 3]
```diff
@@ -25,14 +25,49 @@
...
+    JSOutPoint point;
+    SproutNote note;
+    CAmount amount;
+
+    SendManyInputJSOP(JSOutPoint point_, SproutNote note_, CAmount amount_)
+    {
+        point = point_;
+        note = note_;
+        amount = amount_;
+    }
```
Comment (str4d) [inline src/wallet/asyncrpcoperation_sendmany.h]:
```suggestion
    SendManyInputJSOP(JSOutPoint point_, SproutNote note_, CAmount amount_) :
        point(point_), note(note_), amount(amount_) {}
```

**Comment by oxarbitrage** [comment]
Changes made, thanks @str4d

File: src/wallet/asyncrpcoperation_sendmany.h [hunk 4]
```diff
@@ -25,14 +25,36 @@
...
-typedef std::tuple<std::string, CAmount, std::string> SendManyRecipient;
+class SendManyRecipient {
+public:
+    std::string address;
+    CAmount amount;
+    std::string memo;
 
-// Input UTXO is a tuple (quadruple) of txid, vout, amount, coinbase)
-typedef std::tuple<uint256, int, CAmount, bool> SendManyInputUTXO;
+    SendManyRecipient(std::string address_, CAmount amount_, std::string memo_ = "") :
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.h]: I don't think the optionality of the memo argument is ever used.
```suggestion
    SendManyRecipient(std::string address_, CAmount amount_, std::string memo_) :
```

**Comment by daira** [comment]
@zkbot r+

**Comment by str4d** [comment]
Transient failure in a supported builder.

@zkbot retry

**Pull request merged**